src/mbgl/renderer/renderer_impl.cpp

```diff
@@ -270,6 +236,67 @@ void Renderer::Impl::render(const UpdateParameters& updateParameters) {
         placement->getVariableOffsets()
     };
 
+    Color backgroundColor;
+
+    struct RenderItem {
+        RenderItem(RenderLayer& layer_, RenderSource* source_)
+            : layer(layer_), source(source_) {}
+        RenderLayer& layer;
+        RenderSource* source;
+    };
+
+    std::vector<RenderItem> renderItems;
+    std::vector<const RenderLayerSymbolInterface*> renderItemsWithSymbols;
+
+    // Update all sources and initialize renderItems.
+    renderItems.reserve(layerImpls->size());
+    for (const auto& sourceImpl : *sourceImpls) {
+        RenderSource* source = renderSources.at(sourceImpl->id).get();
+        std::vector<Immutable<Layer::Impl>> filteredLayersForSource;
+        filteredLayersForSource.reserve(layerImpls->size());
+        bool sourceNeedsRendering = false;
+        bool sourceNeedsRelayout = false;       
+
+        for (const auto& layerImpl : *layerImpls) {
+            RenderLayer* layer = getRenderLayer(layerImpl->id);
+            const auto* layerInfo = layerImpl->getTypeInfo();
+            bool layerNeedsRendering = layer->needsRendering(zoomHistory.lastZoom);
+
+            parameters.staticData.has3D = (parameters.staticData.has3D || layerInfo->pass3d == LayerTypeInfo::Pass3D::Required);
+
+            if (layerInfo->source != LayerTypeInfo::Source::NotRequired) {
+                if (layerImpl->source == sourceImpl->id) {
+                    sourceNeedsRendering |= layerNeedsRendering;
+                    sourceNeedsRelayout = (sourceNeedsRelayout || hasImageDiff || hasLayoutDifference(layerDiff, layerImpl->id));
+                    filteredLayersForSource.push_back(layerImpl);
+
+                    if (layerNeedsRendering) {
+                        renderItems.emplace_back(*layer, source);
+                    }
+                }
+                continue;
+            } 
+
+            // Handle layers without source.
+            if (layerNeedsRendering && sourceImpl.get() == sourceImpls->at(0).get()) {            
+                if (parameters.contextMode == GLContextMode::Unique
+                    && layerImpl.get() == layerImpls->at(0).get()) {
+                    const auto& solidBackground = layer->getSolidBackground();
+                    if (solidBackground) {
+                        backgroundColor = *solidBackground;
+                        continue; // This layer is shown with background color, and it shall not be added to render items. 
```

Comment:
nit: maybe worth adding comment for that, for example `see // - CLEAR block`

Reply:
I'm going to update this code shortly with the follow-up patches for #13812 , so I'd like to defer this nit to that moment.